remotes/fixup.go

```diff
@@ -3,86 +3,127 @@ package remotes
+			EventType:      eventType,
+			Message:        message,
+			Error:          err,
+			Progress:       progress.snapshot(),
+		}
+	}
+	notifyImageEndError := func(err error) {
+		notifyEvent(FixupEventTypeCopyImageEnd, "", err)
+	}
 	repoOnly, imageRef, descriptor, err := fixupBaseImage(ctx, &baseImage, ref, resolver)
 	if err != nil {
+		notifyImageEndError(err)
 		return bundle.BaseImage{}, err
 	}
 	if imageRef.Name() == ref.Name() {
+		events <- FixupEvent{
+			DestinationRef: ref,
+			SourceImage:    originalSource,
+			EventType:      FixupEventTypeCopyImageEnd,
+			Message:        "Nothing to do: image reference is already present in repository" + repoOnly.String(),
+		}
 		return baseImage, nil
 	}
-
-	fmt.Fprintln(os.Stderr, "Image is not present in repository")
 	sourceRepoOnly, err := reference.ParseNormalizedNamed(imageRef.Name())
 	if err != nil {
+		notifyImageEndError(err)
 		return bundle.BaseImage{}, err
 	}
 	sourceFetcher, err := resolver.Fetcher(ctx, sourceRepoOnly.Name())
 	if err != nil {
+		notifyImageEndError(err)
 		return bundle.BaseImage{}, err
 	}
 
 	// Prepare the copier or the mounter
-	copier, err := newImageCopier(ctx, resolver, sourceFetcher, repoOnly.String())
+	copier, err := newImageCopier(ctx, resolver, sourceFetcher, repoOnly.String(), notifyEvent)
 	if err != nil {
+		notifyImageEndError(err)
 		return bundle.BaseImage{}, err
 	}
 	var handler imageHandler = &copier
 	if reference.Domain(imageRef) == reference.Domain(ref) {
 		mounter, err := newImageMounter(ctx, resolver, copier, sourceRepoOnly.Name(), repoOnly.Name())
 		if err != nil {
+			notifyImageEndError(err)
 			return bundle.BaseImage{}, err
 		}
 		handler = &mounter
 	}
 
-	// Walk the source repository and list all the descriptors
-	accumulator := &descriptorAccumulator{}
-	if err := images.Walk(ctx, images.Handlers(accumulator, images.ChildrenHandler(&imageContentProvider{sourceFetcher})), descriptor); err != nil {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	wq := newWorkQueue(ctx, 4, 50)
+	walker := newManifestWalker(repoOnly.String(), resolver, images.ChildrenHandler(&imageContentProvider{sourceFetcher}),
+		handler, notifyEvent, wq, progress)
+	if _, err = walker.walk(ctx, descriptor, nil); err != nil {
+		notifyImageEndError(err)
+		wq.stopAndWait()
 		return bundle.BaseImage{}, err
 	}
-	for _, d := range accumulator.descriptors {
-		if err := handler.Handle(ctx, d); err != nil {
-			return bundle.BaseImage{}, err
-		}
+	progress.markWalkDone()
+	notifyEvent(FixupEventTypeProgress, "", nil)
+	if err = wq.stopAndWait(); err != nil {
+		notifyImageEndError(err)
+		return bundle.BaseImage{}, err
 	}
 
+	events <- FixupEvent{
+		DestinationRef: ref,
+		SourceImage:    originalSource,
+		EventType:      FixupEventTypeCopyImageEnd,
+	}
 	return baseImage, nil
 }
 
 func fixupBaseImage(ctx context.Context,
 	baseImage *bundle.BaseImage,
-	ref opts.NamedOption,
+	ref reference.Named, //nolint: interfacer
```

Comment:
What was the `interfacer` complaint here?

Reply:
opts.NamedOption has nothing to do with image references. Even if it satisfies the requirement from the function body, it gives a wrong idea about what the parameter represents.